clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapChainMap.java

```diff
@@ -242,73 +240,49 @@ public void clear() {
     }
   }
 
-  private void evict() {
-    int evictionIndex = heads.getEvictionIndex();
-    if (evictionIndex < 0) {
-      throw new OversizeMappingException("Storage Engine and Eviction Failed - Everything Pinned (" + getSize() + " mappings) \n" + "Storage Engine : " + chainStorage);
-    } else {
-      heads.evict(evictionIndex, false);
-    }
-  }
-
-  private static final Chain EMPTY_CHAIN = new Chain() {
-    @Override
-    public Iterator<Element> reverseIterator() {
-      return Collections.<Element>emptyList().iterator();
-    }
-
-    @Override
-    public boolean isEmpty() {
-      return true;
-    }
-
-    @Override
-    public int length() {
-      return 0;
-    }
-
-    @Override
-    public Iterator<Element> iterator() {
-      return Collections.<Element>emptyList().iterator();
-    }
-  };
-
-  public static Chain chain(ByteBuffer... buffers) {
-    final List<Element> list = new ArrayList<>();
-    for (ByteBuffer b : buffers) {
-      list.add(element(b));
-    }
-
-    return new Chain() {
-
-      final List<Element> elements = Collections.unmodifiableList(list);
-
-      @Override
-      public Iterator<Element> iterator() {
-        return elements.iterator();
-      }
-
-      @Override
-      public Iterator<Element> reverseIterator() {
-        return Util.reverseIterator(elements);
-      }
+  @Override
+  public Iterator<Chain> iterator() {
+    Iterator<Map.Entry<K, InternalChain>> headsIterator = heads.entrySet().iterator();
 
+    return new Iterator<Chain>() {
       @Override
-      public boolean isEmpty() {
-        return elements.isEmpty();
+      public boolean hasNext() {
+        return headsIterator.hasNext();
       }
 
       @Override
-      public int length() {
-        return elements.size();
+      public Chain next() {
+        final Lock lock = heads.readLock();
```

Comment:
I think `org.terracotta.offheapstore.ReadWriteLockedOffHeapClockCache` already takes a lock in `next` (by `org.terracotta.offheapstore.AbstractLockedOffHeapHashMap.LockedEntryIterator#next`).  Do you really need this one as well?

Reply:
It's not safe to drop the lock until after the `detach()` call has been made.